### PR TITLE
[REBASE & FF] Add Various MP Services Related Features

### PR DESCRIPTION
**Pull request by os-d**
## Description

Add a few MP services related commits cherry-picked from 2311 commits:  bece719926
b9ef7cc071
812af0578c

- [x] Impacts functionality?
  - **Functionality** - Does the change ultimately impact how firmware functions?
  - Examples: Add a new library, publish a new PPI, update an algorithm, ...
- [ ] Impacts security?
  - **Security** - Does the change have a direct security impact on an application,
    flow, or firmware?
  - Examples: Crypto algorithm change, buffer overflow fix, parameter
    validation improvement, ...
- [ ] Breaking change?
  - **Breaking change** - Will anyone consuming this change experience a break
    in build or boot behavior?
  - Examples: Add a new library class, move a module to a different repo, call
    a function in a new library class in a pre-existing module, ...
- [ ] Includes tests?
  - **Tests** - Does the change include any explicit test code?
  - Examples: Unit tests, integration tests, robot tests, ...
- [ ] Includes documentation?
  - **Documentation** - Does the change contain explicit documentation additions
    outside direct code modifications (and comments)?
  - Examples: Update readme file, add feature readme file, link to documentation
    on an a separate Web page, ...

## How This Was Tested

From 2311.

## Integration Instructions

N/A.

### CODE REVIEW

File: MdePkg/Include/Ppi/MpServices.h [hunk 1]
```diff
@@ -270,6 +316,7 @@ struct _EFI_PEI_MP_SERVICES_PPI {
   EFI_PEI_MP_SERVICES_SWITCH_BSP                  SwitchBSP;
   EFI_PEI_MP_SERVICES_ENABLEDISABLEAP             EnableDisableAP;
   EFI_PEI_MP_SERVICES_WHOAMI                      WhoAmI;
+  EFI_PEI_MP_SERVICES_STARTUP_THIS_AP_NONBLOCK    StartupThisApNonBlocking;
```
Comment (Javagedes) [inline MdePkg/Include/Ppi/MpServices.h]: Needs MU_CHANGE tag

File: UefiCpuPkg/CpuMpPei/CpuMpPei.c [hunk 2]
```diff
@@ -21,6 +21,7 @@ EFI_PEI_MP_SERVICES_PPI  mMpServicesPpi = {
   PeiSwitchBSP,
   PeiEnableDisableAP,
   PeiWhoAmI,
+  PeiStartupThisAPNonBlocking,
```
Comment (Javagedes) [inline UefiCpuPkg/CpuMpPei/CpuMpPei.c]: Needs MU_CHANGE Tag

File: UefiCpuPkg/Include/Protocol/CpuMpDebug.h [hunk 3]
```diff
@@ -0,0 +1,31 @@
+/** @file -- CpuMpDebug.h
```
Comment (Javagedes) [inline UefiCpuPkg/Include/Protocol/CpuMpDebug.h]: nit: Different header then we normally use, typically just `/** @file`

**Review by Javagedes — APPROVED** [review]

File: UefiCpuPkg/Library/MpInitLib/DxeMpLib.c [hunk 4]
```diff
@@ -34,6 +47,65 @@ UINTN                   mApPageTable;
...
+  UINTN    StackSize,
+  UINTN    CpuNumber,
+  BOOLEAN  IsSwitchStack
+  )
+{
+  CPU_MP_DEBUG_PROTOCOL  *Entry;
+
+  Entry = AllocatePool (sizeof (CPU_MP_DEBUG_PROTOCOL));
+  if (Entry == NULL) {
+    DEBUG ((DEBUG_ERROR, "%a - Failed to allocate memory!\n", __FUNCTION__));
```
Comment (makubacki) [inline UefiCpuPkg/Library/MpInitLib/DxeMpLib.c]: Something to consider.

```suggestion
    DEBUG ((DEBUG_ERROR, "%a - Failed to allocate memory!\n", __func__));
```

File: UefiCpuPkg/Include/Protocol/CpuMpDebug.h [hunk 5]
```diff
@@ -0,0 +1,31 @@
...
+
+This protocol provides debug access to AP buffer information for validation and testing.
+
+Copyright (c) Microsoft Corporation. All rights reserved.
+SPDX-License-Identifier: BSD-2-Clause-Patent
+
+**/
+
+#ifndef __CPU_MP_DEBUG_PROTOCOL__
+#define __CPU_MP_DEBUG_PROTOCOL__
```
Comment (makubacki) [inline UefiCpuPkg/Include/Protocol/CpuMpDebug.h]:
```suggestion
#ifndef CPU_MP_DEBUG_PROTOCOL_
#define CPU_MP_DEBUG_PROTOCOL_
```

File: UefiCpuPkg/Include/Protocol/CpuMpDebug.h [hunk 6]
```diff
@@ -0,0 +1,31 @@
+/** @file -- CpuMpDebug.h
+
+This protocol provides debug access to AP buffer information for validation and testing.
+
+Copyright (c) Microsoft Corporation. All rights reserved.
```
Comment (makubacki) [inline UefiCpuPkg/Include/Protocol/CpuMpDebug.h]:
```suggestion
Copyright (c) Microsoft Corporation.
```

**Review by makubacki — APPROVED** [review]

**Pull request merged**